src/core/qgsvectorlayerupdater.cpp

```diff
@@ -0,0 +1,87 @@
+#include "qgsvectorlayerupdater.h"
+
+
+QgsVectorLayerUpdater::QgsVectorLayerUpdater(
+  QgsVectorLayer &destinationLayer,
+  const QMap<QString, QgsExpression> fieldsMap,
+  QgsExpressionContext &context,
+  const QList<QString> primaryKeys,
+  const bool bypassEditBuffer )
+  : mLayer( destinationLayer )
+  , mFieldsMap( fieldsMap )
+  , mContext( context )
+  , mPrimaryKeys( primaryKeys )
+  , mBypassEditBuffer( bypassEditBuffer )
+{
+
+}
+
+bool QgsVectorLayerUpdater::addFeature( QgsFeature &feature, QgsFeatureSink::Flags flags )
+{
+  Q_UNUSED( flags );
+
+  mContext.setFeature( ( const QgsFeature & ) feature );
+
+  QgsFeature f = QgsFeature();
+  f.setGeometry( feature.geometry() );
+
+  //QgsAttributes &values( mLayer.fields().count() );
+  for ( int i = 0; i < mLayer.fields().count(); i++ )
+  {
+    QgsField field = mLayer.fields()[i];
+    QgsExpression expr = mFieldsMap[field.name()];
+    f.setAttribute( i, expr.evaluate( ( const QgsExpressionContext * ) &mContext ) );
+  }
+
+  bool updated = false;
+  if ( !mPrimaryKeys.isEmpty() )
+  {
+    QgsFeatureRequest request = QgsFeatureRequest();
+    Q_FOREACH ( const QString &pKey, mPrimaryKeys )
+    {
+      request.combineFilterExpression( QgsExpression( QStringLiteral( "%1=%2" )
+                                       .arg( QgsExpression::quotedColumnRef( pKey )
+                                             .arg( QgsExpression::quotedValue( f.attribute( mLayer.fields().indexOf( pKey ) ) ) ) ) ) );
+    }
+    QgsFeatureIterator candidates = mLayer.getFeatures( request );
+    QgsFeature candidate;
+    while ( candidates.nextFeature( candidate ) )
+    {
+      f.setId( candidate.id() );
+      mLayer.updateFeature( f );
```

Comment:
This is rather inefficient -- as the docs for QgsVectorLayer::updateFeature note, it requires another request to the data provider. So as it stands this method is now requesting a subset of features from the layer (usually only resulting in one feature matched), and then updateFeature is firing off a second request to fetch just this feature from the provider. That's going to be a performance killer for large layers. 

I think here you should use QgsVectorLayer::changeAttributeValue and QgsVectorLayer::changeGeometry instead-- those calls will avoid the extra request here.

But in general -- this approach of fetching features one-by-one is always going to be very slow. To fix this you could move all the code in this method from addFeature to addFeatures, and then fetch ALL features to be updated in a single request, and then loop over them updating each in turn using changeAttributeValue & changeGeometry.